File: addin/src/CAD2Revit.Core/Report.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace CAD2Revit.Core
{
    public enum Status
    {
        Placed,     // created (or, in preview, would be created)
        Duplicate,  // an instance already exists at this location
        Unmapped,   // block name not in the mapping file
        Skipped,    // mapped, but family/type not loaded in the project
        Failed,     // Revit refused the placement / no host found
    }

    /// <summary>The outcome for one CAD block (or one unmapped block name).</summary>
    public class PlacementResult
    {
        public string BlockName = "";
        public MapRow Row;
        public Status Status;
        public long? ElementId;
        public string Message = "";
        public double[] Point;          // X, Y, Z in feet, model internal coordinates
        public double? Rotation;        // radians
        public string Host = "";
        public string Level = "";
        public double ScaleX = 1, ScaleY = 1;
        public bool Mirrored;
        public bool HasBlock = true;    // false for grouped "unmapped" rows
        public int Count = 1;
    }

    public class Summary
    {
        public List<KeyValuePair<string, int>> ByType = new List<KeyValuePair<string, int>>();
        public Dictionary<Status, int> Status = new Dictionary<Status, int>();
        public List<KeyValuePair<string, int>> Unmapped = new List<KeyValuePair<string, int>>();
        public List<PlacementResult> Problems = new List<PlacementResult>();
        /// <summary>Placed, but with a "WARNING:" message (e.g. placed level-based instead of on a plane).</summary>
        public List<PlacementResult> Warnings = new List<PlacementResult>();

        public int Get(Status s) => Status.TryGetValue(s, out var n) ? n : 0;
    }

    public static class Report
    {
        const double MmPerFoot = 304.8;
        static readonly CultureInfo Inv = CultureInfo.InvariantCulture;

        public static readonly string[] LogHeader =
        {
            "Status", "CAD_Block", "Family", "Type", "Level", "ElementId", "Host", "X_mm", "Y_mm", "Z_mm",
            "Rotation_deg", "Block_Scale", "Mirrored", "Message",
        };

        public static string StatusText(Status s) => s.ToString().ToLowerInvariant();

        public static Summary Summarize(IEnumerable<PlacementResult> results)
        {
            var s = new Summary();
            var byType = new Dictionary<string, int>();
            var unmapped = new Dictionary<string, int>();
            foreach (var r in results)
            {
                s.Status[r.Status] = s.Get(r.Status) + r.Count;
                if (r.Status == Status.Placed && r.Row != null)
                {
                    byType[r.Row.Label] = (byType.TryGetValue(r.Row.Label, out var n) ? n : 0) + 1;
                    if ((r.Message ?? "").Contains("WARNING:")) s.Warnings.Add(r);
                }
                else if (r.Status == Status.Unmapped)
                    unmapped[r.BlockName] = (unmapped.TryGetValue(r.BlockName, out var m) ? m : 0) + r.Count;
                else if (r.Status == Status.Failed || r.Status == Status.Skipped)
                    s.Problems.Add(r);
            }
            s.ByType = byType.OrderBy(kv => kv.Key, StringComparer.OrdinalIgnoreCase).ToList();
            s.Unmapped = unmapped.OrderBy(kv => kv.Key, StringComparer.OrdinalIgnoreCase).ToList();
            return s;
        }

        /// <summary>Identical problems collapsed: (status, block, message, count).</summary>
        public static List<Tuple<Status, string, string, int>> GroupProblems(IEnumerable<PlacementResult> problems) =>
            problems.GroupBy(r => new { r.Status, r.BlockName, r.Message })
                .Select(g => Tuple.Create(g.Key.Status, g.Key.BlockName, g.Key.Message, g.Count()))
                .OrderBy(t => t.Item1).ThenBy(t => t.Item2, StringComparer.OrdinalIgnoreCase)
                .ToList();

        static string Mm(double v) => (v * MmPerFoot).ToString("0.0", Inv);

        static string Deg(double rad)
        {
            var d = rad * 180.0 / Math.PI % 360.0;
            if (d < 0) d += 360.0;
            if (Math.Abs(d - 360.0) < 0.005) d = 0;
            return d.ToString("0.00", Inv);
        }

        static string G3(double v) => v.ToString("0.###", Inv);

        public static List<IList<object>> LogRows(IEnumerable<PlacementResult> results)
        {
            var rows = new List<IList<object>>();
            foreach (var r in results)
            {
                var p = r.Point;
                string scale = !r.HasBlock ? ""
                    : Math.Abs(r.ScaleX - r.ScaleY) < 1e-6 ? G3(r.ScaleX) : G3(r.ScaleX) + " x " + G3(r.ScaleY);
                rows.Add(new object[]
                {
                    StatusText(r.Status), r.BlockName,
                    r.Row?.Family ?? "", r.Row?.TypeName ?? "", r.Level ?? "",
                    r.ElementId.HasValue ? (object)r.ElementId.Value : "",
                    r.Host ?? "",
                    p != null ? Mm(p[0]) : "", p != null ? Mm(p[1]) : "", p != null ? Mm(p[2]) : "",
                    r.Rotation.HasValue ? Deg(r.Rotation.Value) : "",
                    scale,
                    r.Mirrored ? "yes" : "",
                    r.Count == 1 ? r.Message : $"{r.Count} instance(s). {r.Message}".Trim(),
                });
            }
            return rows;
        }

        /// <summary>Plain-text summary shown in the result window.</summary>
        public static string SummaryText(Summary s, bool preview)
        {
            var lines = new List<string>
            {
                preview ? "PREVIEW - nothing was changed in the model." : "DONE",
                "",
                $"{s.Get(Status.Placed)} {(preview ? "would be placed" : "placed")}, " +
                $"{s.Get(Status.Duplicate)} duplicates skipped, {s.Get(Status.Failed)} failed, " +
                $"{s.Get(Status.Skipped)} not loaded, {s.Get(Status.Unmapped)} unmapped instances",
            };
            if (s.ByType.Count > 0)
            {
                lines.Add("");
                lines.Add(TextTable.Format(new[] { "Family : Type", preview ? "Would place" : "Placed" },
                    s.ByType.Select(kv => new object[] { kv.Key, kv.Value })));
            }
            if (s.Unmapped.Count > 0)
            {
                lines.Add("");
                lines.Add("Unmapped blocks (add them to the mapping file to place them):");
                lines.Add(TextTable.Format(new[] { "CAD block", "Instances" },
                    s.Unmapped.Select(kv => new object[] { kv.Key, kv.Value })));
            }
            if (s.Warnings.Count > 0)
            {
                lines.Add("");
                lines.Add(preview ? "Would be placed, with warnings:" : "Placed with warnings:");
                lines.Add(TextTable.Format(new[] { "CAD block", "Count", "Warning" },
                    s.Warnings.GroupBy(r => new { r.BlockName, r.Message })
                        .Select(g => new object[] { g.Key.BlockName, g.Count(), g.Key.Message })));
            }
            if (s.Problems.Count > 0)
            {
                lines.Add("");
                lines.Add("Failed / skipped:");
                lines.Add(TextTable.Format(new[] { "Status", "CAD block", "Count", "Reason" },
                    GroupProblems(s.Problems).Select(g => new object[] { StatusText(g.Item1), g.Item2, g.Item4, g.Item3 })));
            }
            return string.Join(Environment.NewLine, lines);
        }
    }
}

File: CAD2Revit.extension/lib/cad2revit/report.py
# -*- coding: utf-8 -*-
"""Placement results, summaries and the CSV log. Pure Python (unit tested)."""
import math

MM_PER_FOOT = 304.8

# Result.status values
PLACED = "placed"          # created (or, in preview, would be created)
DUPLICATE = "duplicate"    # an instance already exists at this location
UNMAPPED = "unmapped"      # block name not in the mapping file
SKIPPED = "skipped"        # mapped, but family/type not loaded in the project
FAILED = "failed"          # Revit refused the placement / no host found

LOG_HEADER = ["Status", "CAD_Block", "Family", "Type", "ElementId", "Host",
              "X_mm", "Y_mm", "Z_mm", "Rotation_deg", "Block_Scale", "Mirrored",
              "Message"]


class Result(object):
    def __init__(self, block, row, status, element_id=None, message=u"",
                 point=None, rotation=None, host=u"", count=1):
        self.block = block            # BlockRef, or block name for unmapped
        self.row = row                # MapRow or None
        self.status = status
        self.element_id = element_id  # int or None
        self.message = message
        self.point = point            # (x, y, z) feet, model internal coordinates
        self.rotation = rotation      # radians
        self.host = host              # text, e.g. "Ceilings (linked)"
        self.count = count            # >1 only for grouped "unmapped" rows

    @property
    def block_name(self):
        return getattr(self.block, "name", self.block) or u""


def _xyz(p):
    if p is None:
        return None
    if hasattr(p, "X"):
        return (p.X, p.Y, p.Z)
    return tuple(p)


def _mm(v):
    return u"" if v is None else u"%.1f" % (v * MM_PER_FOOT)


def _deg(rad):
    if rad is None:
        return u""
    d = math.degrees(rad) % 360.0
    return u"%.2f" % (0.0 if abs(d - 360.0) < 0.005 else d)


def summarize(results):
    """Returns dict with:
       by_type:  [(family : type, placed count)] sorted
       status:   {status: count} (unmapped counted per instance)
       unmapped: [(block name, instances)] sorted
       problems: [Result] failed / skipped (not duplicates)"""
    by_type, status, unmapped, problems = {}, {}, {}, []
    for r in results:
        status[r.status] = status.get(r.status, 0) + r.count
        if r.status == PLACED and r.row is not None:
            by_type[r.row.label] = by_type.get(r.row.label, 0) + 1
        elif r.status == UNMAPPED:
            unmapped[r.block_name] = unmapped.get(r.block_name, 0) + r.count
        elif r.status in (FAILED, SKIPPED):
            problems.append(r)
    return {
        "by_type": sorted(by_type.items(), key=lambda kv: kv[0].lower()),
        "status": status,
        "unmapped": sorted(unmapped.items(), key=lambda kv: kv[0].lower()),
        "problems": problems,
    }


def group_problems(problems):
    """[(status, block, message, count)] - identical problems collapsed."""
    groups = {}
    for r in problems:
        k = (r.status, r.block_name, r.message)
        groups[k] = groups.get(k, 0) + 1
    return sorted([k + (n,) for k, n in groups.items()], key=lambda g: (g[0], g[1].lower()))


def log_rows(results):
    """Rows for the CSV log (one per block instance / unmapped block name)."""
    rows = []
    for r in results:
        p = _xyz(r.point)
        b = r.block
        scale = u""
        if hasattr(b, "scale_x"):
            scale = u"%.3g" % b.scale_x if abs(b.scale_x - b.scale_y) < 1e-6 else \
                u"%.3g x %.3g" % (b.scale_x, b.scale_y)
        rows.append([
            r.status, r.block_name,
            r.row.family if r.row else u"", r.row.type_name if r.row else u"",
            r.element_id if r.element_id is not None else u"",
            r.host,
            _mm(p[0]) if p else u"", _mm(p[1]) if p else u"", _mm(p[2]) if p else u"",
            _deg(r.rotation), scale,
            u"yes" if getattr(b, "mirrored", False) else u"",
            r.message if r.count == 1 else u"{} instance(s). {}".format(r.count, r.message).strip(),
        ])
    return rows
